File: 155-min-stack/155-min-stack.py
class MinStack:
    def __init__(self):
        self.s = [] # s[i][0]: value, s[i][1]: cur min

    def push(self, val: int) -> None:
        cur_min = math.inf
        if self.s:
            cur_min = self.s[-1][1]
        self.s.append((val, min(val, cur_min)))

    def pop(self) -> None:
        self.s.pop()
        

    def top(self) -> int:
        return self.s[-1][0]

    def getMin(self) -> int:
        return self.s[-1][1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: turingplanet/stack/155.min-stack.cpp
/*
 * @lc app=leetcode id=155 lang=cpp
 *
 * [155] Min Stack
 */

// @lc code=start
class MinStack
{
  public:
    stack<int> vals;
    stack<int> mins;

    MinStack() {}

    void push(int val)
    {
        vals.push(val);
        // push to mins if val is <= mins.top()
        // making it a monotonic stack
        if (mins.empty() || mins.top() >= val)
            mins.push(val);
    }

    void pop()
    {
        int v = vals.top();
        if (!mins.empty() && mins.top() == v)
            mins.pop();
        vals.pop();
    }

    int top() { return vals.top(); }

    int getMin() { return mins.top(); }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */
// @lc code=end
